File: tests/TgLLM.Integration.Tests/SendKeyboardPlanExpirySingleUseTests.fs
/// Acceptance tests for `TgBot.SendKeyboardPlan`'s send-time `expiresIn`/`singleUse` options: they
/// stamp every tool binding the send produces with an `ExpiresAt`/`SingleUse` derived from the
/// bot's own clock, and an expired-by-send-time binding is refused like an unknown token — same
/// "ack, no invoke" contract `LifecycleTests.fs` already proves for a binding mutated directly in
/// the store, but exercised here through the public send-time knobs instead.
module TgLLM.Integration.Tests.SendKeyboardPlanExpirySingleUseTests

open System
open System.Threading
open System.Threading.Tasks
open System.Text.Json.Nodes
open Expecto
open FSharp.UMX
open TgLLM.Core
open TgLLM.FSharp
open TgLLM.Integration.Tests.FakeBotApiServer

let private field (key: string) (node: JsonNode) : JsonNode =
    match node.[key] |> Option.ofObj with
    | Some c -> c
    | None -> failwithf "expected JSON field '%s'" key

let private at (i: int) (node: JsonNode) : JsonNode =
    match node.[i] |> Option.ofObj with
    | Some c -> c
    | None -> failwithf "expected JSON index %d" i

let private asString (node: JsonNode) : string = node.AsValue().GetValue<string>()

let private callbackDataAt (row: int) (col: int) (sendBody: JsonNode) : string =
    sendBody |> field "reply_markup" |> field "inline_keyboard" |> at row |> at col |> field "callback_data" |> asString

let private planOrFail (rows: PlanButton list list) : ToolKeyboard =
    match Plan.rows rows with
    | Ok p -> p
    | Error e -> failtestf "plan should be valid: %A" e

let private parseToken (tokenStr: string) : CallbackToken =
    match CallbackToken.tryParse tokenStr with
    | ValueSome t -> t
    | ValueNone -> failtest "the sent keyboard's callback_data must parse as a canonical token"

let private pollUntil (ms: int) (predicate: unit -> bool) : Task =
    task {
        let mutable tries = 0

        while not (predicate ()) && tries < ms / 10 do
            do! Task.Delay 10
            tries <- tries + 1

        if not (predicate ()) then failtest "timed out waiting for the expected condition"
    }

let private config (server: FakeBotApiServer) (tools: ToolRegistry) (store: IBindingStore) (clock: Clock) : TgBotConfig =
    (TgBotConfig.create "123456789:TEST-fake-token")
        .WithBaseUrl(server.BaseUrl)
        .WithTools(tools)
        .WithBindingStore(store)
        .WithClock(clock)

[<Tests>]
let sendKeyboardPlanExpirySingleUseTests =
    testList
        "TgBot.SendKeyboardPlan expiresIn/singleUse"
        [

          testCaseAsync "a keyboard sent with expiresIn stores bindings whose ExpiresAt is the bot's clock plus the given span"
          <| async {
              do!
                  task {
                      use! server = FakeBotApiServer.start ()
                      let store = InMemoryBindingStore() :> IBindingStore
                      let now = DateTimeOffset.UtcNow
                      let tools = ToolRegistry.create().Register("approve", fun _ -> task { return () })
                      use! bot = TgBot.startPolling (config server tools store (fun () -> now))

                      let plan = planOrFail [ [ Plan.tool "Approve" "approve" ] ]

                      let! _ =
                          bot.SendKeyboardPlan(
                              UMX.tag<chatId> 9101L,
                              MessageText.unsafe "Deploy?",
                              plan,
                              expiresIn = TimeSpan.FromMinutes 10.0
                          )

                      let sentKeyboard = (server.RequestsFor "sendMessage").Head.Body |> Option.get
                      let token = parseToken (callbackDataAt 0 0 sentKeyboard)
                      let! stored = store.TryGet(token, CancellationToken.None)

                      match stored with
                      | ValueSome binding ->
                          Expect.equal binding.ExpiresAt (Some(now.AddMinutes 10.0)) "ExpiresAt is the bot's clock plus expiresIn"
                      | ValueNone -> failtest "the freshly-sent keyboard's binding must already be in the store"
                  }
                  |> Async.AwaitTask
          }

          testCaseAsync "a keyboard sent with singleUse = true stores bindings stamped as single-use"
          <| async {
              do!
                  task {
                      use! server = FakeBotApiServer.start ()
                      let store = InMemoryBindingStore() :> IBindingStore
                      let tools = ToolRegistry.create().Register("confirm", fun _ -> task { return () })
                      use! bot = TgBot.startPolling ((TgBotConfig.create "123456789:TEST-fake-token").WithBaseUrl(server.BaseUrl).WithTools(tools).WithBindingStore(store))

                      let plan = planOrFail [ [ Plan.tool "Confirm" "confirm" ] ]
                      let! _ = bot.SendKeyboardPlan(UMX.tag<chatId> 9102L, MessageText.unsafe "Confirm?", plan, singleUse = true)

                      let sentKeyboard = (server.RequestsFor "sendMessage").Head.Body |> Option.get
                      let token = parseToken (callbackDataAt 0 0 sentKeyboard)
                      let! stored = store.TryGet(token, CancellationToken.None)

                      match stored with
                      | ValueSome binding -> Expect.isTrue binding.SingleUse "SingleUse is stamped onto every binding this send produces"
                      | ValueNone -> failtest "the freshly-sent keyboard's binding must already be in the store"
                  }
                  |> Async.AwaitTask
          }

          testCaseAsync "a tap on a keyboard sent with a short expiresIn is refused once the bot's clock has advanced past it"
          <| async {
              do!
                  task {
                      use! server = FakeBotApiServer.start ()
                      let store = InMemoryBindingStore() :> IBindingStore
                      let mutable now = DateTimeOffset.UtcNow
                      let toolRan = TaskCompletionSource()
                      let tools = ToolRegistry.create().Register("approve", fun _ -> task { toolRan.TrySetResult() |> ignore })
                      let chat = 9103L
                      use! bot = TgBot.startPolling (config server tools store (fun () -> now))

                      let plan = planOrFail [ [ Plan.tool "Approve" "approve" ] ]

                      let! _ =
                          bot.SendKeyboardPlan(
                              UMX.tag<chatId> chat,
                              MessageText.unsafe "Deploy?",
                              plan,
                              expiresIn = TimeSpan.FromSeconds 30.0
                          )

                      let sentKeyboard = (server.RequestsFor "sendMessage").Head.Body |> Option.get
                      let tokenStr = callbackDataAt 0 0 sentKeyboard

                      // Advance the bot's own clock well past the binding's send-time expiry —
                      // deterministic, no real sleep needed.
                      now <- now.AddMinutes 5.0

                      server.EnqueueResult(
                          "getUpdates",
                          TelegramJson.batch [ TelegramJson.callbackQueryUpdate 1 "q-send-time-expiry" tokenStr chat 60 921L "Presser" ]
                      )

                      do! pollUntil 5000 (fun () -> server.RequestsFor "answerCallbackQuery" |> List.isEmpty |> not)

                      Expect.isFalse toolRan.Task.IsCompleted "a tap after the send-time expiry has elapsed must never run the tool"
                  }
                  |> Async.AwaitTask
          } ]

File: tests/TgLLM.CSharp.Tests/SendKeyboardPlanExpirySingleUseTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.FSharp.Collections;
using TgLLM.CSharp;
using Xunit;
using FakeServerModule = TgLLM.Integration.Tests.FakeBotApiServer.FakeBotApiServerModule;
using TelegramJson = TgLLM.Integration.Tests.FakeBotApiServer.TelegramJson;

namespace TgLLM.CSharp.Tests;

/// <summary>
/// Façade tests for <see cref="TelegramAgent.SendKeyboardPlanAsync"/>'s <c>expiresIn</c>/
/// <c>singleUse</c> parameters: they stamp every tool binding the send produces, and a single-use
/// binding stops routing after its first successful press — driven entirely through the C# façade,
/// reusing the fake Bot API server the F# integration suite already proves against real
/// Telegram.Bot request/response shapes.
/// </summary>
public class SendKeyboardPlanExpirySingleUseTests
{
    private const string FakeToken = "123456789:TEST-fake-token";

    /// <summary>A minimal in-memory <see cref="IBindingStoreCSharp"/> that lets a test inspect a
    /// binding's stamped fields directly, the same shape as any C# host's own store.</summary>
    private sealed class InMemoryCSharpStore : IBindingStoreCSharp
    {
        private readonly Dictionary<string, ToolBindingDto> _bindings = new();

        public ValueTask SaveAsync(IReadOnlyList<ToolBindingDto> bindings, CancellationToken ct)
        {
            foreach (var binding in bindings)
            {
                _bindings[binding.Token] = binding;
            }

            return ValueTask.CompletedTask;
        }

        public ValueTask<ToolBindingDto?> TryGetAsync(string token, CancellationToken ct) =>
            ValueTask.FromResult(_bindings.TryGetValue(token, out var binding) ? binding : null);

        public ValueTask RemoveAsync(IReadOnlyList<string> tokens, CancellationToken ct)
        {
            foreach (var token in tokens)
            {
                _bindings.Remove(token);
            }

            return ValueTask.CompletedTask;
        }

        public ValueTask<int> EvictExpiredAsync(DateTimeOffset now, CancellationToken ct) => ValueTask.FromResult(0);
    }

    private static string CallbackDataAt(JsonNode sendBody, int row, int col) =>
        sendBody["reply_markup"]!["inline_keyboard"]![row]![col]!["callback_data"]!.GetValue<string>();

    private static FSharpList<string> ListOf(params string[] items) => ListModule.OfArray(items);

    private static async Task WaitUntilAsync(Func<bool> predicate, int timeoutMs, CancellationToken ct)
    {
        var deadline = DateTime.UtcNow.AddMilliseconds(timeoutMs);

        while (!predicate())
        {
            if (DateTime.UtcNow > deadline)
            {
                throw new TimeoutException("timed out waiting for the expected condition");
            }

            await Task.Delay(10, ct);
        }
    }

    [Fact]
    public async Task Sending_a_keyboard_with_expiresIn_and_singleUse_stamps_both_onto_its_binding()
    {
        var ct = TestContext.Current.CancellationToken;
        await using var server = await FakeServerModule.start();
        const long chatId = 4401L;

        var tools = new ToolRegistry().Register("approve", _ => Task.CompletedTask);
        var csharpStore = new InMemoryCSharpStore();
        var store = BindingStoreAdapter.ToCoreStore(csharpStore);

        await using var agent = await TelegramAgent.StartPollingAsync(
            new TelegramAgentOptions { BotToken = FakeToken, BaseUrl = server.BaseUrl, Tools = tools, BindingStore = store },
            ct);

        var plan = new PlanBuilder().Row(r => r.Tool("Approve", "approve")).Build();
        var beforeSend = DateTimeOffset.UtcNow;

        await agent.SendKeyboardPlanAsync(chatId, "Deploy?", plan, expiresIn: TimeSpan.FromMinutes(10), singleUse: true, ct: ct);

        var sentKeyboard = server.RequestsFor("sendMessage").First().Body!.Value;
        var token = CallbackDataAt(sentKeyboard, 0, 0);

        var stored = await csharpStore.TryGetAsync(token, ct);

        Assert.NotNull(stored);
        Assert.NotNull(stored!.ExpiresAt);
        Assert.True(stored.ExpiresAt!.Value >= beforeSend.AddMinutes(10), "expiresIn should be stamped at least ten minutes past the send");
        Assert.True(stored.ExpiresAt!.Value <= DateTimeOffset.UtcNow.AddMinutes(10).AddSeconds(5), "expiresIn should be stamped roughly ten minutes past the send, not further out");
        Assert.True(stored.SingleUse, "singleUse should be stamped onto the binding");
    }

    [Fact]
    public async Task A_keyboard_sent_with_singleUse_true_runs_its_tool_once_then_ignores_a_second_tap()
    {
        var ct = TestContext.Current.CancellationToken;
        await using var server = await FakeServerModule.start();
        const long chatId = 4402L;
        var runCount = 0;

        var tools = new ToolRegistry().Register(
            "confirm",
            _ =>
            {
                Interlocked.Increment(ref runCount);
                return Task.CompletedTask;
            });

        await using var agent = await TelegramAgent.StartPollingAsync(
            new TelegramAgentOptions { BotToken = FakeToken, BaseUrl = server.BaseUrl, Tools = tools },
            ct);

        var plan = new PlanBuilder().Row(r => r.Tool("Confirm", "confirm")).Build();
        await agent.SendKeyboardPlanAsync(chatId, "Confirm?", plan, singleUse: true, ct: ct);

        var sentKeyboard = server.RequestsFor("sendMessage").First().Body!.Value;
        var token = CallbackDataAt(sentKeyboard, 0, 0);

        server.EnqueueResult(
            "getUpdates",
            TelegramJson.batch(ListOf(TelegramJson.callbackQueryUpdate(1, "q-single-use-first", token, chatId, 70, 960L, "Presser"))));

        await WaitUntilAsync(() => Volatile.Read(ref runCount) >= 1, 5000, ct);
        Assert.Equal(1, Volatile.Read(ref runCount));

        server.EnqueueResult(
            "getUpdates",
            TelegramJson.batch(ListOf(TelegramJson.callbackQueryUpdate(2, "q-single-use-second", token, chatId, 71, 960L, "Presser"))));

        await WaitUntilAsync(() => server.RequestsFor("answerCallbackQuery").Count() >= 2, 5000, ct);
        Assert.Equal(1, Volatile.Read(ref runCount));
    }
}
